File: app/copilot/copilot-backend/src/main/java/com/microsoft/openai/samples/assistant/config/agent/TransactionAgent.java
package com.microsoft.openai.samples.assistant.config.agent;


import com.google.adk.agents.LlmAgent;
import com.google.adk.models.langchain4j.LangChain4j;
import com.google.adk.tools.AgentTool;
import com.google.adk.tools.BaseTool;
import com.google.adk.tools.FunctionTool;
import com.google.adk.tools.mcp.McpTool;
import dev.langchain4j.model.chat.ChatModel;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.util.ArrayList;
import java.util.List;

@Configuration
public class TransactionAgent {
    String transactionMCPServerUrl;
    ChatModel chatModel;
    LlmAgent accountAgent;

    public TransactionAgent(ChatModel chatModel,
                            @Value("${transactions.api.url}") String transactionMCPServerUrl,
                            @Qualifier("adkAccountAgent") LlmAgent accountAgent){
        this.chatModel  = chatModel;
        this.transactionMCPServerUrl = transactionMCPServerUrl;
        this.accountAgent = accountAgent;
    }

    @Bean(name = "adkTransactionAgent")
    public LlmAgent getAgent() {

        List<McpTool> transactionHistoryTools = ADKUtils.buildMCPTools(this.transactionMCPServerUrl);

        //Build a BaseTool list out of MCP tools and add the InvoiceScanTool
        List<BaseTool> allTools = new ArrayList<>(transactionHistoryTools);
        allTools.add(AgentTool.create(accountAgent));

        return LlmAgent.builder()
                .name("TransactionAgent")
                .model(new LangChain4j(this.chatModel))
                .description("Agent to help users review their account transactions, banking movements and payments history.")
                .instruction("""
                        you are a personal financial advisor who help the users to review their account transactions, banking movements and payments history.
                        
                        ## 1. Business Rules & Guardrails
                            - **Privacy First:**
                              Do NOT display, store, or process any full account numbers, Social Security Numbers, card CVV codes, or other sensitive personally identifiable information beyond the user's partial account name or masked number (e.g., ****4321).
                            - **No Actionable Changes:**
                              Never access, alter, or initiate payments, transfers, or any other account changes. This assistant is strictly informational.
                            - **Data Scope:**
                              - If the user want to search last transactions for a specific payee, ask to provide the payee name
                              - Only reference transactions, balances, and payment history within the past 3 months.
                              - Only show the last 10 transactions ordered by date by default.
                            - **Error Handling:**
                              If input is incomplete, ambiguous, or not understood, politely request clarification without guessing about financial information.
                        ## 2. User Query Handling & Response Formatting
                            - **Acceptable Queries:**
                              - “Show me my last transactions.”
                              - “Can I see payments made to Amazon in May?”
                              - “List all direct debits from my checking account.”
                              - “What was my highest payment last month?”
                              - "When was last time I paid my electricity bill?"
                        
                            - **Formatting Rules:**
                              - List relevant transactions in a clear html tabular format with columns such as ID, Date, Description, Payment Type, Amount, Recipient Name.
                              - Always include a summary above the table (e.g., “Here are your 10 most recent payments to Amazon”) and a note reminding about privacy (e.g., “Note: Sensitive account details are masked for your security.”). \s
                             
                        
                        ## 3. Use below context information
                              - Current Timestamp: '{timestamp}'
                              - Logged User name: '{loggedUserName}'
                    """)
                .tools(allTools)
                .build();
    }


}

File: app/copilot/app/agents/azure_chat/transaction_agent.py
from agent_framework.azure import AzureChatClient
from agent_framework import ChatAgent, MCPStreamableHTTPTool
from datetime import datetime

import logging


logger = logging.getLogger(__name__)

class TransactionHistoryAgent :
    instructions = """
    you are a personal financial advisor who help the user with their recurrent bill payments. To search about the payments history you need to know the payee name.
    By default you should search the last 10 account transactions ordered by date.    
    If the user want to search last account transactions for a specific payee, extract it from the request and use it as filter.
    
    Use html list or table to display the transaction information.
    Always use the below logged user details to retrieve account info:
    {user_mail}
    Current timestamp:
    {current_date_time}
    """
    name = "TransactionHistoryAgent"
    description = "This agent manages user transactions related information such as banking movements and payments history"

    def __init__(self, azure_chat_client: AzureChatClient,
                 account_mcp_server_url: str,
                 transaction_mcp_server_url: str,
                  ):
        self.azure_chat_client = azure_chat_client
        self.account_mcp_server_url = account_mcp_server_url
        self.transaction_mcp_server_url = transaction_mcp_server_url
      


    async def build_af_agent(self) -> ChatAgent:
    
      logger.info("Building request scoped transaction agent run ")
      
      user_mail="bob.user@contoso.com"
      current_date_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
      full_instruction = TransactionHistoryAgent.instructions.format(user_mail=user_mail, current_date_time=current_date_time)

      logger.info("Initializing Account MCP server tools ")
      #await self.account_mcp_server.__aenter__()
      account_mcp_server = MCPStreamableHTTPTool(
        name="Account MCP server client",
        url=self.account_mcp_server_url
     )
      await account_mcp_server.connect()
     
      logger.info("Initializing Transaction MCP server tools ")
      transaction_mcp_server = MCPStreamableHTTPTool(
        name="Transaction MCP server client",
        url=self.transaction_mcp_server_url
     )
      await transaction_mcp_server.connect()

      return self.azure_chat_client.create_agent(
           instructions=full_instruction,
           name=TransactionHistoryAgent.name,
           tools=[account_mcp_server, transaction_mcp_server])